Purchases/Purchasing/StoreKit2/ProductsFetcherSK1.swift

```diff
@@ -0,0 +1,127 @@
+//
+//  Licensed under the MIT License (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      https://opensource.org/licenses/MIT
+//
+//  ProductsFetcherSK1.swift
+//
+//  Created by Andrés Boedo on 7/26/21.
+
+import Foundation
+import StoreKit
+
+class ProductsFetcherSK1: NSObject {
+    private let productsRequestFactory: ProductsRequestFactory
+
+    private var cachedProductsByIdentifier: [String: SKProduct] = [:]
+    private let queue = DispatchQueue(label: "ProductsFetcherSK1")
+    private var productsByRequests: [SKRequest: Set<String>] = [:]
+    private var completionHandlers: [Set<String>: [(Set<SKProduct>) -> Void]] = [:]
+
+    init(productsRequestFactory: ProductsRequestFactory = ProductsRequestFactory()) {
+        self.productsRequestFactory = productsRequestFactory
+    }
+
+    func products(withIdentifiers identifiers: Set<String>,
+                  completion: @escaping (Set<SKProduct>) -> Void) {
+
+        queue.async { [self] in
+            let productsAlreadyCached = self.cachedProductsByIdentifier.filter { key, _ in identifiers.contains(key) }
+            if productsAlreadyCached.count == identifiers.count {
+                let productsAlreadyCachedSet = Set(productsAlreadyCached.values)
+                Logger.debug(String(format: Strings.offering.products_already_cached, identifiers))
+                completion(productsAlreadyCachedSet)
+                return
+            }
+
+            if let existingHandlers = self.completionHandlers[identifiers] {
+                Logger.debug(String(format: Strings.offering.found_existing_product_request, identifiers))
+                self.completionHandlers[identifiers] = existingHandlers + [completion]
+                return
+            }
+
+            Logger.debug(
+                String(format: Strings.offering.no_cached_requests_and_products_starting_skproduct_request, identifiers)
+            )
+            let request = self.productsRequestFactory.request(productIdentifiers: identifiers)
+            request.delegate = self
+            self.completionHandlers[identifiers] = [completion]
+            self.productsByRequests[request] = identifiers
+            request.start()
+        }
+    }
+}
+
+extension ProductsFetcherSK1: SKProductsRequestDelegate {
+
+    func productsRequest(_ request: SKProductsRequest, didReceive response: SKProductsResponse) {
+        queue.async { [self] in
+            Logger.rcSuccess(Strings.network.skproductsrequest_received_response)
+            guard let requestProducts = self.productsByRequests[request] else {
+                Logger.error("requested products not found for request: \(request)")
+                return
+            }
+            guard let completionBlocks = self.completionHandlers[requestProducts] else {
+                Logger.error("callback not found for failing request: \(request)")
+                return
```

Comment:
would it make sense to call `self.productsByRequests.removeValue(forKey: request)` here?

Reply:
yeah, no reason not to. This would only happen in the case of a programming error, though. We never wanted to do a `fatalError` for it, but perhaps we should (just a thought for somewhere in the future)